Blinky example won't start when changing blinky.c -> blinky.cpp

**Describe the bug**
C++ version of main will compile fine, but LED won't start flashing.

**Sources to reproduce**
I think it's so simple I won't provide a fork, but I can if it's really needed.

**Steps to reproduce**
1. Go to Blinky example project directory
2. Rename blinky.c -> blinky.cpp
2. Change the source file in CMakeLists.txt
3. Add CXX as the project language in CMakeLists.txt
4. Compile and flash your microcontroller (I'm using USB DFU if that makes a difference)
5. LED is not flashing as it should

**Expected behavior**
LED should flash as in the C example. Nothing will happen with C++ version. Something wrong with startup procedure on C++?

**Environment (please complete the following information):**
 - OS: Fedora 34
 - Compiler: arm-none-eabi-gcc (Fedora 10.2.0-4.fc34) 10.2.0
 - stm32-cmake: commit 12240e0
 - cmake: 3.19.7
 - HAL/cube/CMSIS: 3d7a3c1 Release v2.6.6, 313c279 Release v1.7.12, 96d6da4 Official ARM version: v5.6.0

If I'll add `STM32::Nano` target for linker (as guided in the docs), it will give me linker errors:
```
[ 87%] Linking CXX executable executable.elf
/usr/lib/gcc/arm-none-eabi/10.2.0/../../../../arm-none-eabi/bin/ld: /usr/lib/gcc/arm-none-eabi/10.2.0/../../../../arm-none-eabi/lib/thumb/v7e-m+fp/hard/libc_nano.a(lib_a-abort.o): in function `abort':
/builddir/build/BUILD/newlib-4.1.0/build-nano/arm-none-eabi/thumb/v7e-m+fp/hard/newlib/libc/stdlib/../../../../../../../../newlib/libc/stdlib/abort.c:59: undefined reference to `_exit'
/usr/lib/gcc/arm-none-eabi/10.2.0/../../../../arm-none-eabi/bin/ld: /usr/lib/gcc/arm-none-eabi/10.2.0/../../../../arm-none-eabi/lib/thumb/v7e-m+fp/hard/libc_nano.a(lib_a-signalr.o): in function `_kill_r':
/builddir/build/BUILD/newlib-4.1.0/build-nano/arm-none-eabi/thumb/v7e-m+fp/hard/newlib/libc/reent/../../../../../../../../newlib/libc/reent/signalr.c:53: undefined reference to `_kill'
/usr/lib/gcc/arm-none-eabi/10.2.0/../../../../arm-none-eabi/bin/ld: /usr/lib/gcc/arm-none-eabi/10.2.0/../../../../arm-none-eabi/lib/thumb/v7e-m+fp/hard/libc_nano.a(lib_a-signalr.o): in function `_getpid_r':
/builddir/build/BUILD/newlib-4.1.0/build-nano/arm-none-eabi/thumb/v7e-m+fp/hard/newlib/libc/reent/../../../../../../../../newlib/libc/reent/signalr.c:83: undefined reference to `_getpid'
collect2: error: ld returned 1 exit status
```

--- a/README.md
+++ b/README.md
@@ -32,7 +32,7 @@ It uses cmake and GCC, along with newlib (libc), STM32Cube. Supports F0 F1 F2 F3
 * `blinky` ([examples/blinky](examples/blinky)) - blink led using STM32 HAL library and SysTick.
    It will compile a project for the `F4` family by default, but you can also compile for the
    `L0` and `F1` family by passing `BLINKY_L0_EXAMPLE=ON` or `BLINKY_F1_EXAMPLE=ON` to the CMake
-   generation call.
+   generation call. Using C++ instead of C is possible using `USE_CPP_FILE=ON`.
 * `freertos` ([examples/freertos](examples/freertos)) - blink led using STM32 HAL library and FreeRTOS.
    You need to specify at least one board by passing `FREERTOS_<BOARD>_EXAMPLE=ON` to CMake.
    Currently, the example can be built for the `H743ZI` and `F407VG` board targets.

--- a/examples/blinky/CMakeLists.txt
+++ b/examples/blinky/CMakeLists.txt
@@ -1,7 +1,15 @@
 cmake_minimum_required(VERSION 3.16)
 set(CMAKE_TOOLCHAIN_FILE ${CMAKE_CURRENT_SOURCE_DIR}/../../cmake/stm32_gcc.cmake)
 
-project(stm32-blinky C ASM)
+option(USE_CPP_FILE "Use the blink.cpp file instead of the blink.c" OFF)
+
+if(USE_CPP_FILE)
+    project(stm32-blinky C CXX ASM)
+    set(MAIN_SOURCE_FILE blinky.cpp)
+else()
+    project(stm32-blinky C ASM)
+    set(MAIN_SOURCE_FILE blinky.c)
+endif()
 set(CMAKE_INCLUDE_CURRENT_DIR TRUE)
 
 # Configure here which STM32 target(s) to build
@@ -41,7 +49,7 @@ find_package(HAL COMPONENTS "${HAL_COMP_LIST}" REQUIRED)
 
 # STM32F4-Discovery
 if(BLINKY_F4_EXAMPLE)
-    add_executable(stm32-blinky-f4 blinky.c stm32f4xx_hal_conf.h)
+    add_executable(stm32-blinky-f4 ${MAIN_SOURCE_FILE} stm32f4xx_hal_conf.h)
     target_link_libraries(stm32-blinky-f4
         HAL::STM32::F4::RCC
         HAL::STM32::F4::GPIO
@@ -54,7 +62,7 @@ endif()
 
 # STM32VL-Discovery
 if(BLINKY_F1_EXAMPLE)
-    add_executable(stm32-blinky-f1 blinky.c stm32f1xx_hal_conf.h)
+    add_executable(stm32-blinky-f1 ${MAIN_SOURCE_FILE} stm32f1xx_hal_conf.h)
     target_link_libraries(stm32-blinky-f1
         HAL::STM32::F1::RCC
         HAL::STM32::F1::GPIO
@@ -67,7 +75,7 @@ endif()
 
 # STM32L0538-Discovery
 if(BLINKY_L0_EXAMPLE)
-    add_executable(stm32-blinky-l0 blinky.c stm32l0xx_hal_conf.h)
+    add_executable(stm32-blinky-l0 ${MAIN_SOURCE_FILE} stm32l0xx_hal_conf.h)
     target_link_libraries(stm32-blinky-l0
         HAL::STM32::L0::RCC
         HAL::STM32::L0::GPIO

--- a/examples/blinky/CMakeLists.txt
+++ b/examples/blinky/CMakeLists.txt
@@ -1,7 +1,12 @@
 cmake_minimum_required(VERSION 3.16)
 set(CMAKE_TOOLCHAIN_FILE ${CMAKE_CURRENT_SOURCE_DIR}/../../cmake/stm32_gcc.cmake)
 
-option(USE_CPP_FILE "Use the blink.cpp file instead of the blink.c" OFF)
+# Configure here which STM32 target(s) to build
+option(BLINKY_F4_EXAMPLE "Compile F4 example" ON)
+option(BLINKY_F1_EXAMPLE "Compile F1 example" OFF)
+option(BLINKY_L0_EXAMPLE "Compile L0 example" OFF)
+
+option(USE_CPP_FILE "Use the .cpp example file instead of the .c one" OFF)
 
 if(USE_CPP_FILE)
     project(stm32-blinky C CXX ASM)
@@ -10,12 +15,9 @@ else()
     project(stm32-blinky C ASM)
     set(MAIN_SOURCE_FILE blinky.c)
 endif()
+
 set(CMAKE_INCLUDE_CURRENT_DIR TRUE)
 
-# Configure here which STM32 target(s) to build
-option(BLINKY_F4_EXAMPLE "Compile F4 example" ON)
-option(BLINKY_F1_EXAMPLE "Compile F1 example" OFF)
-option(BLINKY_L0_EXAMPLE "Compile L0 example" OFF)
 
 set(HAL_COMP_LIST RCC GPIO CORTEX)
 set(CMSIS_COMP_LIST "")

--- a/examples/blinky/blinky.c
+++ b/examples/blinky/blinky.c
@@ -27,7 +27,9 @@ void SysTick_Handler(void)
 
     // 1 Hz blinking
     if ((HAL_GetTick() % 500) == 0)
+    {
         HAL_GPIO_TogglePin(LED_PORT, LED_PIN);
+    }
 }
 
 void initGPIO()

--- a/examples/blinky/blinky.cpp
+++ b/examples/blinky/blinky.cpp
@@ -30,7 +30,9 @@ extern "C"
 
         // 1 Hz blinking
         if ((HAL_GetTick() % 500) == 0)
+        {
             HAL_GPIO_TogglePin(LED_PORT, LED_PIN);
+        }
     }
 }
 

--- a/examples/blinky/blinky.cpp
+++ b/examples/blinky/blinky.cpp
@@ -1,0 +1,62 @@
+#if defined STM32L0
+    #include <stm32l0xx_hal.h>
+
+    // STM32L0538-Discovery green led - PB4
+    #define LED_PORT                GPIOB
+    #define LED_PIN                 GPIO_PIN_4
+    #define LED_PORT_CLK_ENABLE     __HAL_RCC_GPIOB_CLK_ENABLE
+#elif defined STM32F1
+    #include <stm32f1xx_hal.h>
+
+    // STM32VL-Discovery green led - PC9
+    #define LED_PORT                GPIOC
+    #define LED_PIN                 GPIO_PIN_9
+    #define LED_PORT_CLK_ENABLE     __HAL_RCC_GPIOC_CLK_ENABLE
+#elif defined STM32F4
+    #include <stm32f4xx_hal.h>
+
+    // STM32F4-Discovery green led - PD12
+    #define LED_PORT                GPIOD
+    #define LED_PIN                 GPIO_PIN_12
+    #define LED_PORT_CLK_ENABLE     __HAL_RCC_GPIOD_CLK_ENABLE
+#endif
+
+//This prevent name mangling for functions used in C/assembly files.
+extern "C"
+{
+    void SysTick_Handler(void)
+    {
+        HAL_IncTick();
+
+        // 1 Hz blinking
+        if ((HAL_GetTick() % 500) == 0)
+            HAL_GPIO_TogglePin(LED_PORT, LED_PIN);
+    }
+}
+
+void initGPIO()
+{
+    GPIO_InitTypeDef GPIO_Config;
+
+    GPIO_Config.Mode = GPIO_MODE_OUTPUT_PP;
+    GPIO_Config.Pull = GPIO_NOPULL;
+    GPIO_Config.Speed = GPIO_SPEED_FREQ_HIGH;
+
+    GPIO_Config.Pin = LED_PIN;
+
+    LED_PORT_CLK_ENABLE();
+    HAL_GPIO_Init(LED_PORT, &GPIO_Config);
+}
+
+int main(void)
+{
+    HAL_Init();
+    initGPIO();
+    // 1kHz ticks
+    HAL_SYSTICK_Config(SystemCoreClock / 1000);
+
+    for (;;)
+        __WFI();
+
+    return 0;
+}

